### PR TITLE
[SPARK-35335][SQL] Coalesce shuffle partition as much as possible for REPARTITION_BY_NONE

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Set minPartitionNum to 1 for `CoalesceShufflePartitions` to coalesce shuffle partition as much as possible for `REPARTITION_BY_NONE`.

### Why are the changes needed?

Before this PR. We use `repartition` to avoid generating small files. But it still generates `spark.sql.adaptive.coalescePartitions.minPartitionNum` number of files, which may leads to exceeding the directory item limitation:
```
org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.hdfs.protocol.FSLimitException$MaxDirectoryItemsExceededException): The directory item limit of /path/to/table is exceeded: limit=1048576 items=1048576
```

After this PR. We can avoid generating small files:
```sql
INSERT INTO TABLE target_tbl SELECT /*+ REPARTITION */ * FROM t
```
```scala
df.repartition().write.saveAsTable("target_tbl")
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @maryannxue FYI

**Comment by ulysses-you** [comment]
Thank you for introducing this idea. But that's not only about `spark.sql.adaptive.coalescePartitions.minPartitionNum ` but also other adaptive configs, such as `spark.sql.adaptive.coalescePartitions.enabled`, `spark.sql.adaptive.advisoryPartitionSizeInBytes`.

Follow this thought, we can split stages into two parts to make configs isolated with each other:
* final stage (for write)
* previous stage (for query)

And the code may look like:
```
withSQLConf("final stage config") {
  applyPhysicalRules(
        result.newPlan,
        finalStageOptimizerRules,
        Some((planChangeLogger, "AQE Final Query Stage Optimization")))
}
```

The thought is not very mature but it's quite useful in our scenario.

**Comment by ulysses-you** [comment]
Some random thoughts.

We considered about supporting stage level config completly that means for every query stage we can use it's own config. Some options:
* enhance hint to support set config and at AQE side use it. Syntax may look like `SELECT /*+ MERGE(t2, config(key=value)) */ * FROM t1 JOIN t2 JOIN t3`
* add config tag to match the stage id.  Syntax may look like `set spark.sql.adaptive.stage1.advisoryPartitionSizeInBytes=xxx`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala [hunk 1]
```diff
@@ -83,10 +84,16 @@ case class CoalesceShufflePartitions(session: SparkSession) extends CustomShuffl
         // is not set, so to avoid perf regressions compared to no coalescing.
         val minPartitionNum = conf.getConf(SQLConf.COALESCE_PARTITIONS_MIN_PARTITION_NUM)
           .getOrElse(session.sparkContext.defaultParallelism)
+        val minNumPartitions = if (isFinalStage) {
```
Comment (AngersZhuuuu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala:87]: minFinalStagePartitionNum？

**Pull request closed without merging**